File: 프로그래머스/lv2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    #그래프 초기화
    graph = []
    cnt = 1
    for i in range(rows):
        List = []
        for j in range(columns):
            List.append(cnt)
            cnt += 1
        graph.append(List)

    answer = []

    for querie in queries:
        x1, y1, x2, y2 = querie
        x1, y1, x2, y2 = x1-1, y1-1, x2-1, y2-1

        Min = int(1e9)

        new_value = dict()

        #오른쪽 이동 처리
        for i in range(y1,y2):
            new_value[(x1,i+1)] = graph[x1][i]
            Min = min(Min, graph[x1][i])

        #아래로 이동 처리
        for i in range(x1,x2):
            new_value[(i+1,y2)] = graph[i][y2]
            Min = min(Min, graph[i][y2])

        #왼쪽으로 이동 처리
        for i in range(y2,y1,-1):
            new_value[(x2,i-1)] = graph[x2][i]
            Min = min(Min, graph[x2][i])

        #위로 이동 처리
        for i in range(x2,x1,-1):
            new_value[(i-1,y1)] = graph[i][y1]
            Min = min(Min, graph[i][y1])

        #변경사항 그래프에 적용
        for key in new_value:
            x, y = key
            graph[x][y] = new_value[key]

        answer.append(Min)
    return answer

File: 프로그래머스/lv2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
//11:00
//테두리 부분을 시계방향으로 회전, 

import java.util.*;

class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];
        
        //graph 초기화
        int count = 1;
        int[][] graph = new int[rows][columns];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                graph[i][j] = count;
                count++;
            }
        }
        
        int ans_index = 0;
        
        //쿼리 실행
        for (int[] query : queries) {
            
            //인덱스 정리
            int x1 = query[0]-1, y1 = query[1]-1, x2 = query[2]-1, y2 = query[3]-1;
            
            int value = 100000000;
            
            //큐 초기화
            Deque<Integer> queue = new LinkedList<Integer>();
            for (int i = y1; i <= y2; i++) {
                queue.add(graph[x1][i]);
                value = Math.min(value, graph[x1][i]);
            }
            
            for (int i = x1+1; i < x2; i++) {
                queue.add(graph[i][y2]);
                value = Math.min(value, graph[i][y2]);
            }
            
            for (int i = y2; i >= y1; i--) {
                queue.add(graph[x2][i]);
                value = Math.min(value, graph[x2][i]);
            }
            
            for (int i = x2-1; i > x1; i--) {
                queue.add(graph[i][y1]);
                value = Math.min(value, graph[i][y1]);
            }
            
            
            queue.addFirst(queue.pollLast());

            //배열 초기화
            for (int i = y1; i <= y2; i++) {
                graph[x1][i] = queue.pollFirst();
            }
            
            for (int i = x1+1; i < x2; i++) {
                graph[i][y2] = queue.pollFirst();
            }
            
            for (int i = y2; i >= y1; i--) {
                graph[x2][i] = queue.pollFirst();
            }
            
            for (int i = x2-1; i > x1; i--) {
                graph[i][y1] = queue.pollFirst();
            }
            
            answer[ans_index] = value;
            ans_index++;
        }
        
        return answer;
    }
}